### ISSUE TITLE
Add options to lasso that improve performance

### ISSUE THREAD

**Issue by rmweir** [rancher/rancher#37350]
**Is your feature request related to a problem? Please describe.**
Certain default behaviors of lasso cause resource consumption spikes and apiserver request spikes. These behavior are related to informer cache resyncs.

**Describe the solution you'd like**
* There should be an option to only run update handler on informer if the object has been update. This will be off by default. This aims to disable no change resyncs for applications that have existing mechanisms for handling these scenarios while also keeping resyncs in the case that events are missed. This is a particularly good solution for rancher where separating controller run is not enough as one or two types can have many objects.
* A user should be able to easily customize sync intervals for a given type. If not provided then a sensible default is used.

**Describe alternatives you've considered**
Automatically offset the sync time for each controllers.
There is no builtin way to do this, this can be achieve by the application using lasso by just starting controllers at different times.

Set resync default to 0.
Offers no insurance in the case of a missed event if that were to ever happen.

Reduce workers.
Doesn’t appear to work since a lot of memory isn’t instantly released anyway.

Enqueue objects using AddRateLimited in lasso.
This uses a bucket rated limited with a burst of 100 and a refill rate of 10/s. This isn’t one size fits all when it comes to scale. This will be a good or bad fit depending on setup size.

**Comment by rmweir** [comment]
**Testing:**
For QA:
Enabled metrics with controllers
Provision downstream cluster and deploy 10 projects in it with a few mysql apps deployed in each
Can create 10 projects in local cluster (Attempting to creating enough objects where controller spikes are noticeable)
Set CATTLE_RESYNC_DEFAULT rancher env var to 5
confirm that controllers are running every 5 minutes
Set CATTLE_SKIP_NO_CHANGE_UPDATE to mgmt
Confirm handlers spikes every 5 minutes are smaller
Set CATTLE_SKIP_NO_CHANGE_UPDATE to mgmt,user
Should be even smaller
Set CATTLE_SKIP_NO_CHANGE_UPDATE to mgmt,user,scaled
Will probably look comparable or smaller.

For scaling QA:
Provision a scaled setup, something comparable to a benchmark setup, like ~500 clusters
Enable controllers metrics
Set CATTLE_RESYNC_DEFAULT rancher env var to 20
Confirm controller spikes happen on 20 min interval
See if there is an observable memory spike around when controllers spike.
If so, see if Set CATTLE_SKIP_NO_CHANGE_UPDATE set to mgmt appears to reduce it
do the same as above with env var set to mgmt,user

**Comment by git-ival** [comment]
> **Testing:** For QA: Enabled metrics with controllers Provision downstream cluster and deploy 10 projects in it with a few mysql apps deployed in each Can create 10 projects in local cluster (Attempting to creating enough objects where controller spikes are noticeable) Set CATTLE_RESYNC_DEFAULT rancher env var to 5 confirm that controllers are running every 5 minutes Set CATTLE_SKIP_NO_CHANGE_UPDATE to mgmt Confirm handlers spikes every 5 minutes are smaller Set CATTLE_SKIP_NO_CHANGE_UPDATE to mgmt,user Should be even smaller Set CATTLE_SKIP_NO_CHANGE_UPDATE to mgmt,user,scaled Will probably look comparable or smaller.
> 
> For scaling QA: Provision a scaled setup, something comparable to a benchmark setup, like ~500 clusters Enable controllers metrics Set CATTLE_RESYNC_DEFAULT rancher env var to 20 Confirm controller spikes happen on 20 min interval See if there is an observable memory spike around when controllers spike. If so, see if Set CATTLE_SKIP_NO_CHANGE_UPDATE set to mgmt appears to reduce it do the same as above with env var set to mgmt,user

@rmweir Just to clarify: Should the CATTLE_SKIP_NO_CHANGE_UPDATE env var be set/updated on the existing setup or should a new setup be created for each time we want to set CATTLE_SKIP_NO_CHANGE_UPDATE?

**Comment by git-ival** [comment]
For testing against rancher 2.5-head/2.5.14 we will need to install the latest rancher 2.5 version of rancher-monitoring then setup and install the old controller metrics dashboard as detailed here: https://github.com/rancher/rancher/wiki/How-to-setup-prometheus-dashboard-for-controllers

**Comment by git-ival** [comment]
> > **Testing:** For QA: Enabled metrics with controllers Provision downstream cluster and deploy 10 projects in it with a few mysql apps deployed in each Can create 10 projects in local cluster (Attempting to creating enough objects where controller spikes are noticeable) Set CATTLE_RESYNC_DEFAULT rancher env var to 5 confirm that controllers are running every 5 minutes Set CATTLE_SKIP_NO_CHANGE_UPDATE to mgmt Confirm handlers spikes every 5 minutes are smaller Set CATTLE_SKIP_NO_CHANGE_UPDATE to mgmt,user Should be even smaller Set CATTLE_SKIP_NO_CHANGE_UPDATE to mgmt,user,scaled Will probably look comparable or smaller.
> > For scaling QA: Provision a scaled setup, something comparable to a benchmark setup, like ~500 clusters Enable controllers metrics Set CATTLE_RESYNC_DEFAULT rancher env var to 20 Confirm controller spikes happen on 20 min interval See if there is an observable memory spike around when controllers spike. If so, see if Set CATTLE_SKIP_NO_CHANGE_UPDATE set to mgmt appears to reduce it do the same as above with env var set to mgmt,user
> 
> @rmweir Just to clarify:
> 
> 1. Should the CATTLE_SKIP_NO_CHANGE_UPDATE env var be set/updated on the existing setup or should a new setup be created for each time we want to set CATTLE_SKIP_NO_CHANGE_UPDATE?
> 2. Does it matter what k8s distribution is used for the Local cluster (RKE1, K3s, RKE2)?
> 3. I assume downstreams being imported K3s is ok?

Answered during our regular meeting:
1. The env var can be updated during the same scale test, no new setup needed.
2. This should not matter.
3. This should be fine for testing purposes.

**Comment by rmweir** [comment]
@git-ival The env var has been changed from from CATTLE_SKIP_NO_CHANGE_UPDATE to CATTLE_SYNC_ONLY_CHANGED_OBJECTS.
1. Updated should be fine.
2. No.
3. Yes.

**Comment by jbilliau-rcd** [comment]
@rmweir We just upgraded our nonprod environment last night from 2.5.12 to 2.5.14, which has this fix. Figured I'd give you some metrics/data so you can see what real-world impact of the change. **TL;DR - Memory usage looks way better!**

### 2.5.12:

<img width="425" alt="image" src="https://user-images.githubusercontent.com/69006307/173191719-bc7c5481-330b-4a7f-9181-d4d452180876.png">

You can see the leader pod in particular continually ramps up memory usage more and more over time until it exhausts the worker node and fails, causing a disconnect of all child clusters connected to it (not a great experience for our users)

![image](https://user-images.githubusercontent.com/69006307/173191753-b52e6115-54ea-4797-abc7-91cb74f3cec5.png)

### 2.5.4: (so far, we 12 hours in or so)

During the Helm upgrade command, we passed in these arguments:

--set 'extraEnv[0].name=CATTLE_SYNC_ONLY_CHANGED_OBJECTS' --set 'extraEnv[0].value=user\,mgmt' \
--set 'extraEnv[1].name=CATTLE_RESYNC_DEFAULT' --set-string extraEnv[1].value=20

![image](https://user-images.githubusercontent.com/69006307/173191804-a57fce2d-4324-4ca4-8b35-c7c08afb3d9b.png)

Much lower, more stable, and not continually ramping up. Monday will be the true test as we'll have hundreds of users actively using Rancher, but so far so good!

My question to you is; are there any implications of having a lower resync of 20 vs the default of 600? Also, we are only syncing the user and mgmt contexts now....what other contexts are there? Does it matter they arent being synced now? Basically, I'm wondering if there aren't any cons to this approach, why isn't it just the default?

Thanks!

**Comment by rmweir** [comment]
@jbilliau-rcd the improvement here should be from the `CATTLE_SYNC_ONLY_CHANGED_OBJECTS` env var. The `CATTLE_RESYNC_DEFAULT` env var was only set in my instruction for QA to be able to test the impact without having a long lived setup. You should be able to remove `CATTLE_RESYNC_DEFAULT` and it will go back to the default of 15 hours or so.

**Comment by rmweir** [comment]
Testing this forward port will be very similar to the v2.5 version. The old grafana dashboard is gone so instead we'll want to use two graphs in the Performance-Debugging dashboard. What I did to test it myself is the follow:
1. Install monitoring in rancher setup with some amount of objects installed. I created some project locally and had a single cluster installed I believe. I recommend installing more and probably having more cluster specifically. The more loaded the setup the more obvious changes will be.
2. Change `CATTLE_RESYNC_DEFAULT` env var to 5, then observe a few resyncs in the `Lasso Handler Executions` graphs and the `Handler Executions Over Last 2 Minutes` graph.
3. Set `CATTLE_SYNC_ONLY_CHANGED_OBJECTS` env var to "mgmt", then observe a few resyncs.
4. Set `CATTLE_SYNC_ONLY_CHANGED_OBJECTS` env var to "mgmt,user", then observe a few resyncs.
5. 4. Set `CATTLE_SYNC_ONLY_CHANGED_OBJECTS` env var to "mgmt,user,scaled", then observe a few resyncs.

This is what my graphs looked like:
![image](https://user-images.githubusercontent.com/19376037/174196033-068d4899-5106-444c-a946-5eb442706d39.png)

Exact behavior will vary depending on the type, location (cluster-wise), and number of objects in setup.

With the value mgmt you can see some bumps disappear. With "mgmt,user" more bumps disapear to where there are almost no resync bumps. The resync bump left at this point in my graph is from the `tls-storage` handler for secrets. This can be disregarded as it's a handler that is created and ran by dynamic listener.

One difference to note from v2.6 and v2.5 is that some user handlers have been broken out into "agent handlers". These handlers for downstream clusters are now ran by the downstream clusters themselves and are apart of the clusters "mgmt" context. Some of what used to be user controllers may now be controlled by "mgmt" but only for the local cluster. This will result in possibly seeing some of these clusters run in much smaller amounts than before and are toggled by "mgmt" instead of "user".

**Comment by jbilliau-rcd** [comment]
Thanks for all the detail @rmweir , this is what I was looking for. So is there any reason to ever NOT use `CATTLE_SYNC_ONLY_CHANGED_OBJECTS`? Being as this issue only cropped up for us on 2.5.12, how has memory issue not been prevalent in every release prior due to `CATTLE_SYNC_ONLY_CHANGED_OBJECTS` only being a new env var in 2.5.14?

Also, what exactly is "syncing"? The pods with each other, the pods with etcd, or what?

Lastly, we still running with `CATTLE_RESYNC_DEFAULT` of 20, so 20 minutes if I understand this correctly. Should we keep that? If we got rid of it, and only synced every 15 hours, what are the ramifications of that? Hence, my "what exactly is syncing?" question above :)

**Comment by rmweir** [comment]
@jbilliau-rcd I would just unset `CATTLE_RESYNC_DEFAULT`. The syncing we are talking about here has to do with the cache informers. Since we register handlers to these caches, resyncing them causes all the handlers to rerun. This occurs whether there was a change or not. Since we register most controllers when rancher starts, this means xn runs of handlers where x is the number of objects and n is the number of handlers. It is more nuanced than that but I think it's close enough.

Resyncing the informers is done for two reasons. First, resyncing the caches just in case an event was missed; it ensures the objects we have in memory are accurate. Second, rerunning handlers that may be due to run again even if an object hasn't changed; sometimes this is necessary if the handlers depends on some sort of outside service or ran into an error.

`CATTLE_SYNC_ONLY_CHANGED_OBJECTS` basically prevents handlers from running if the object hasn't changed. Our assertion is that we are not dependent on resyncing for the second objective mentioned above because we have other builtin mechanisms in our controller/handler framework that rerun handlers when needed. Therefore, we shouldn't need to run them every x hours. Even if we do, x hours is likely not an acceptable amount of time to wait for a correction. The main risk is that if we do have some sort of bug in one of our handlers that we are not covering with one of these mechanisms, toggling this removes your safety blanket.

As far as why are things worse than before? I can only give our theories here as there are multiple parts:

* We had (and have) two sets of handlers that ran on two different frameworks. Now, they use the same underlying framework which may have caused them to sync much closer to one another timewise.
* You may have an increase in some objects such as rolebindings.
* New handlers and features have been added.

**Comment by git-ival** [comment]
@rmweir LGTM: I tested on a Small-scaled Rancher cluster with Rancher version `2.6.6` and k3s version `1.22.11+k3s1`. 

Here are some observations:
- Scaled up to max documented limit with:
  - Less than 71% memory utilization.
  - Less than 65% CPU utilization.
    - CPU utilization spiked to near 100% at multiple points.
- Resource utilization and controller/handler frequecy seems to spike up to/above the max at-scale amount when:
  - CATTLE_RESYNC_DEFAULT env var was set to 20.
  - CATTLE_SYNC_ONLY_CHANGED_OBJECTS env var was set to mgmt.
  - CATTLE_SYNC_ONLY_CHANGED_OBJECTS env var was set to mgmt,user.
- Notes: 
  - Upon initially setting each env var the CPU utilization spikes up to 100%. 
  - Memory utilization only ever seems to surpass the max at scale utilization by no more than roughly 8%.
  - During the re-sync intervals (every 20 minutes):
  -  The Resource utilization and controller handler frequency spikes are drastically reduced, although still significant (reaching 5 million or more executions depending on the controller).
  - After setting CATTLE_SYNC_ONLY_CHANGED_OBJECTS the re-sync spikes on controller executions were non-existent as there was no activity within the cluster post scale-in period. Resource utilization still spiked slightly during resync intervals but this seems expected.
  - The most-executed Lasso handler is the `secret-deferred` handler which is executed 4x more than the following 2 most-executed handlers, `monitoring-deferred` and `cluster-crtb-deferred`

Below are some relevant metrics graphs. The first 2 or so hours correlate to the scale-in period of cluster provisioning, followed by a roughly 30 minute idle period. Finally, the relevant environment variables were set, with 20 minute idle periods following the rancher pods restarting.

API Server Requests:
![prometheus_api_server_requests](https://user-images.githubusercontent.com/85186645/178361501-7a97e43b-2d4a-422e-9674-6b9a9272d4f4.png)

API Server 409s:
![prometheus_api_server_409s](https://user-images.githubusercontent.com/85186645/178361620-50193196-d3cc-475e-8a5b-3350b9862345.png)

CPU Utilization:
![lasso_cpu_utilization](https://user-images.githubusercontent.com/85186645/178361734-00e352b3-23e2-438f-8d45-f9195ded14db.png)

Memory Utilization:
![lasso_memory_utilization](https://user-images.githubusercontent.com/85186645/178361872-1e8b8d3d-6c3b-4ca4-90a7-10ee51fa9c59.png)

Lasso Handler Executions:
![lasso_handler_executions](https://user-images.githubusercontent.com/85186645/178361930-619db886-9d82-46ca-8fb8-88d873e918c4.png)

**Comment by OmnesPotens** [comment]
Hi @rmweir what was the reasoning for this being reopened and what needs testing here? I was under the impression that this was closed in 2.6.6

**Comment by brudnak** [comment]
# ✅ PASSED

## Reproduction Environment

#### Not required.

<hr>

## Validation Environment

| Component                        | Version / Type                                                                     |
|----------------------------------|------------------------------------------------------------------------------------|
| Rancher version                  | v2.6-c7996108b0332e7f2a13bd20d5992aad80dcf74b-head                                 |
| Rancher commit link              | https://github.com/rancher/rancher/commit/c7996108b0332e7f2a13bd20d5992aad80dcf74b |
| Installation option              | Helm (high availability)                                                           |
| If Helm Chart k8s cluster        | v1.24.4+k3s1                                                                       |
| Cert Details                     | Let's Encrypt / `traefik`                                                          |
| k3s ha external db               | Aurora MySQL 5.7.mysql_aurora.2.10.2                                               |
| Helm version                     | v2.16.8-rancher2                                                                   |
| Downstream cluster type          | RKE1 Linode                                                                        |
| Downstream K8s version           | v1.24.4-rancher1-1                                                                 |
| Authentication providers enabled | local                                                                              |
| Logged in user role              | Administrator                                                                      |
| Browser type                     | Google Chrome                                                                      |
| Browser version                  | 105.0.5195.125 (Official Build) (x86_64)                                           |

### ADDITIONAL VALIDATION SETUP

```shell
helm install rancher rancher-latest/rancher \
  --namespace cattle-system \
  --set hostname=$URL_VAR \
  --set bootstrapPassword=$PASSWORD_VAR \
  --set ingress.tls.source=letsEncrypt \
  --set letsEncrypt.email=$EMAIL_VAR \
  --set letsEncrypt.ingress.class=traefik \
  --set rancherImageTag=v2.6-head
```

## Validation steps

1. Enabled metrics with controllers
2. Provision downstream cluster and deploy 10 projects in it with a few mysql apps deployed in each
3. Can create 10 projects in local cluster (Attempting to creating enough objects where controller spikes are noticeable)
4. Set CATTLE_RESYNC_DEFAULT rancher env var to 5
5. confirm that controllers are running every 5 minutes
6. Set CATTLE_SKIP_NO_CHANGE_UPDATE to mgmt
7. Confirm handlers spikes every 5 minutes are smaller
8. Set CATTLE_SKIP_NO_CHANGE_UPDATE to mgmt,user
9. Should be even smaller
10. Set CATTLE_SKIP_NO_CHANGE_UPDATE to mgmt,user,scaled
11. Will probably look comparable or smaller.

![2022-09-28_23-21-39](https://user-images.githubusercontent.com/40877149/192957995-1aa242a1-e542-49d3-b3f2-6218b4b1ddf6.png)


## Additional Info

## RESULTS

#### ✅ Expected
Handler spikes smaller

#### ✅ Actual
Handler spikes were smaller
